src/lib.rs

```diff
@@ -1785,6 +1785,139 @@ fn collect_runs(
+) -> PyResult<Vec<Vec<PyObject>>> {
+    let mut pending_list: Vec<Vec<PyObject>> = Vec::new();
+    let mut block_id: Vec<Option<usize>> = Vec::new();
+    let mut block_list: Vec<Vec<PyObject>> = Vec::new();
+
+    let filter_node = |node: &PyObject| -> PyResult<Option<bool>> {
+        let res = filter_fn.call1(py, (node,))?;
+        res.extract(py)
+    };
+
+    let color_edge = |edge: &PyObject| -> PyResult<Option<usize>> {
+        let res = color_fn.call1(py, (edge,))?;
+        res.extract(py)
+    };
+
+    let nodes = match algo::toposort(graph, None) {
+        Ok(nodes) => nodes,
+        Err(_err) => {
+            return Err(DAGHasCycle::new_err("Sort encountered a cycle"))
+        }
+    };
+
+    // Utility for ensuring pending_list has the color index
+    macro_rules! ensure_vector_has_index {
+        ($pending_list: expr, $block_id: expr, $color: expr) => {
+            if $color >= $pending_list.len() {
+                $pending_list.resize($color + 1, Vec::new());
+                $block_id.resize($color + 1, None);
+            }
+        };
+    }
+
+    for node in nodes {
+        if let Some(is_match) = filter_node(&graph.graph[node])? {
+            let raw_edges = graph
+                .graph
+                .edges_directed(node, petgraph::Direction::Outgoing);
+
+            // Remove all edges that do not yield errors from color_fn
+            let colors = raw_edges
+                .map(|edge| {
+                    let edge_weight = edge.weight();
+                    color_edge(edge_weight)
+                })
+                .collect::<PyResult<Vec<Option<usize>>>>()?;
+
+            // Remove null edges from color_fn
+            let colors = colors.into_iter().flatten().collect::<Vec<usize>>();
+
+            if colors.len() <= 2 && is_match {
+                if colors.len() == 1 {
+                    let c0 = colors[0];
+                    ensure_vector_has_index!(pending_list, block_id, c0);
+                    if let Some(c0_block_id) = block_id[c0] {
+                        block_list[c0_block_id]
+                            .push(graph.graph[node].clone_ref(py));
+                    } else {
+                        pending_list[c0].push(graph.graph[node].clone_ref(py));
+                    }
+                } else if colors.len() == 2 {
+                    let c0 = colors[0];
+                    let c1 = colors[1];
+                    ensure_vector_has_index!(pending_list, block_id, c0);
+                    ensure_vector_has_index!(pending_list, block_id, c1);
+
+                    if block_id[c0].is_some()
+                        && block_id[c1].is_some()
+                        && block_id[c0] == block_id[c1]
+                    {
+                        block_list[block_id[c0].unwrap_or_default()]
+                            .push(graph.graph[node].clone_ref(py));
+                    } else {
+                        let mut new_block: Vec<PyObject> = Vec::new();
```

Comment:
Couldn't this be something like:

```suggestion
                        let mut new_block: Vec<PyObject> = Vec::with_capacity(pending_list[c0].len() + pending_list[c1].len() + 1);
```

to avoid the extra allocations?